### PR TITLE
Added support for cursor hiding.

### PR DESCRIPTION
**Pull request by petrikvladimir**

### CODE REVIEW

**Comment by fschutt** [comment]
Well, this shouldn't be a boolean field, you should rather modify `mouse_cursor_type` to be an `Option<Cursor>`, and then check `if mouse_cursor_type.is_none()`. Second, you forgot the code to show the cursor again - so yes, you can hide it once, but then it would be indefinitely hidden. So I'd change those two things.

File: azul/src/window.rs [hunk 1]
```diff
@@ -1061,6 +1061,11 @@ impl<'a, T: Layout> Window<T> {
             window.set_cursor(new_state.mouse_state.mouse_cursor_type);
             old_state.mouse_state.mouse_cursor_type = new_state.mouse_state.mouse_cursor_type;
         }
+        
+        if old_state.is_cursor_hidden != new_state.is_cursor_hidden {
+            old_state.is_cursor_hidden = new_state.is_cursor_hidden;
+            window.hide_cursor(old_state.is_cursor_hidden);
```
Comment (fschutt) [inline azul/src/window.rs]: should be `new_state.is_cursor_hidden`

**Comment by voop** [comment]
Thanks for the comments. It makes sense. However, I was not able to modify mouse cursor properties from the application. For window state modifying in a callback function I am using e.g.: 
```
    if let Some(window) = app_state.windows.get_mut(event.window_id) {
            window.state.is_cursor_hidden = !window.state.is_cursor_hidden;
    }
```
Is that correct?
Do you have suggestions on how to modify the cursor properties in a callback?
Function
```
pub(crate) fn set_mouse_state(...) 
```
is private outside the crate.

**Comment by fschutt** [comment]
You are not supposed to call `set_mouse_state` outside of the crate - the cursor state is set after all the callbacks are run. This is because when there are multiple callbacks that change the cursor state called directly after each other, the cursor would simply flicker - so the cursor type is set once-per-frame instead of once-per-callback.

The mouse state is applied here: https://github.com/maps4print/azul/blob/a98a5b0f4e3c3ad23c32c5fd011f3f06d6eaa87b/azul/src/app.rs#L608-L611

and here: https://github.com/maps4print/azul/blob/a98a5b0f4e3c3ad23c32c5fd011f3f06d6eaa87b/azul/src/app.rs#L700-L703

The problem is probably the second call to `set_mouse_state`, which always resets the state to the default mouse cursor. That should probably be refactored. Remember, that the window you get in the callbacks isn't the real, OS-level window, you only get access to the windows state, not the window itself.

As an additional note, you can use `?` in callbacks:

```rust
let window = app_state.windows.get_mut(event.window_id)?;
window.state.is_cursor_hidden = !window.state.is_cursor_hidden;
```

**Comment by voop** [comment]
Ok, therefore I have to change the state of the window. However, the mouse state in the window state is not public too: `pub(crate) mouse_state: MouseState,`. I guess that it is because the position/button states are updated from window events before the callbacks processing, and therefore the changing that state in the callback is not allowed. Setting mouse_state to default ensures that one mouse click is not processed twice.

Would one of these solutions seems reasonable to you?:
1) In MouseState make all fields except MouseCursor public to create only. 
    In WindowState make field mouse_state public.
    When setting the mouse state to default (after callbacks are done) use the mouse cursor updated in a callback (i.e. will not be overwritten by default value).

2) Add additional MouseCursor state into WindowState which will be then used to change the cursor.

**Comment by fschutt** [comment]
I'm closing this because cursor hiding is now implemented in `synchronize_mouse_state()`.

**Pull request closed without merging**